book/src/dev/rfcs/0010-v5-transaction.md

```diff
@@ -0,0 +1,240 @@
+    orchard_anchor: tree::Root,
+    orchard_shielded_data: Option<OrchardShieldedData>,
+}
+```
+
+`SaplingShieldedDataV5` is different from `SaplingShieldedDataV4` so a new type will be defined and implemented in `zebra-chain/src/transaction/sapling_shielded_data.rs`.
+Definition will look as follows:
+
+```
+pub struct SaplingShieldedDataV5 {
+    pub first: Either<SpendV5, OutputV5>,
+    pub rest_spends: Vec<SpendV5>,
+    pub rest_outputs: Vec<OutputV5>,
+    pub binding_sig: Signature<Binding>,
+}
+```
+
+This leads to new sapling output and spend types.
+
+V5 sapling spend will be defined and implemented in `zebra-chain/src/sapling/spend.rs` (this is in the same file as `SpendV4`):
+```
+pub struct SpendV5 {
+    pub cv: commitment::ValueCommitment,
+    pub nullifier: note::Nullifier,
+    pub rk: redjubjub::VerificationKeyBytes<SpendAuth>,
+    pub zkproof: Groth16Proof,
+    pub spend_auth_sig: redjubjub::Signature<SpendAuth>,
+}
+```
+
+V5 sapling output will live in `zebra-chain/src/sapling/output.rs` (this is in the same file as `OutputV4`).
+
+Note: the V4 and V5 sapling outputs currently have identical fields. We use different types to consistently distinguish V4 and V5 transaction data.
+
+```
+pub struct OutputV5 {
+    pub cv: commitment::ValueCommitment,
+    pub cm_u: jubjub::Fq,
+    pub ephemeral_key: keys::EphemeralPublicKey,
+    pub enc_ciphertext: note::EncryptedNote,
+    pub out_ciphertext: note::WrappedNoteKey,
+    pub zkproof: Groth16Proof,
+}
+```
+To abstract over the `V4` and `V5` `Transaction` enum variants, we will implement `Transaction` methods to access sapling and orchard data. 
+
+To abstract over the `SaplingShieldedDataV4`/`SaplingShieldedDataV5`, `SpendV4`/`SpendV5`, and `OutputV4`/`OutputV5` structs, we will implement `SaplingShieldedData`, `Spend`, and `Output` traits to access sapling data.
+
+The methods in these traits will be implemented in a similar way to the existing `Transaction` methods.
+The new V5 structure will create a new `OrchardShieldedData` type. This new type will be defined in a separated file: `zebra-chain/src/transaction/orchard_shielded_data.rs` and it will look as follows:
+```
+pub struct OrchardShieldedData {
+    /// An action description.
+    ///
+    /// Storing this separately ensures that it is impossible to construct
+    /// an invalid `OrchardShieldedData` with no actions.
+    pub first: Action,
+    pub rest: Vec<Action>,
+    pub binding_sig: Signature<Binding>,
+}
+```
+
+Where `Action` is defined as [Action definition](https://github.com/ZcashFoundation/zebra/blob/68c12d045b63ed49dd1963dd2dc22eb991f3998c/zebra-chain/src/orchard/action.rs#L18-L41).
+
+Finally, in the V5 transaction we have a new `OrchardFlags`. This is a bitfield type defined as:
+
+```
+bitflags! {
+    pub struct OrchardFlags: u8 {
+        const ENABLE_SPENDS = 0b00000001;
+        const ENABLE_OUTPUTS = 0b00000010;
+        // Reserved, zeros (bits 2 .. 7)
+    }
+}
+```
+
```

Comment:
Here's the design update for #1917:

```suggestion

## Security

To avoid parsing memory exhaustion attacks, we will make the following changes across all `Transaction`, `ShieldedData`, `Spend` and `Output` variants, including `V1` to `V4`:
- `Box` large `Optional` data structures, to defer allocation until the variant or option is actually used
- Statically assert that these data structures remain as small as possible
- Check cardinality consensus rules at parse time, before deserializing any `Vec`s
  - In general, Zcash requires that each transaction has at least one Transparent/Sprout/Sapling/Orchard transfer, this rule is not currently encoded in our data structures
- Stop parsing as soon as the first error is detected

See ticket [#1917](https://github.com/ZcashFoundation/zebra/issues/1917) for details.

```